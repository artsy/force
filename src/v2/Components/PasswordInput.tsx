import { ClosedEyeIcon, OpenEyeIcon, space } from "@artsy/palette"
import { useState } from "react"
import * as React from "react"
import styled from "styled-components"
import QuickInput, { QuickInputProps } from "./QuickInput"

export interface PasswordInputProps extends QuickInputProps {
  showPasswordMessage?: boolean
}

/**
 * Password input.
 * Renders the label inside of the textbox; shows/hides the password.
 * @deprecated: Use @artsy/palette's PasswordInput instead.
 */
export const PasswordInput: React.FC<PasswordInputProps> = ({
  error,
  showPasswordMessage,
  ref,
  ...newProps
}) => {
  const [showPassword, toggleShowPassword] = useState(false)

  const eyeIcon = () => {
    const icon = showPassword ? (
      <ClosedEyeIcon onClick={() => toggleShowPassword(false)} />
    ) : (
      <OpenEyeIcon onClick={() => toggleShowPassword(true)} />
    )

    return <Eye>{icon}</Eye>
  }

  const type = showPassword ? "text" : "password"
  const note =
    !error && showPasswordMessage && "Password must be at least 8 characters."

  return (
    <QuickInput
      {...newProps}
      error={error}
      type={type}
      rightAddOn={eyeIcon()}
      // @ts-expect-error PLEASE_FIX_ME_STRICT_NULL_CHECK_MIGRATION
      note={note}
    />
  )
}

const Eye = styled.span`
  position: absolute;
  right: ${space(1)}px;
  z-index: 1;
`
