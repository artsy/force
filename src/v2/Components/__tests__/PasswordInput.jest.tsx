import { mount } from "enzyme"
import "jest-styled-components"
import renderer from "react-test-renderer"

import { PasswordInput } from "../PasswordInput"

describe("PasswordInput", () => {
  it("renders a PasswordInput", () => {
    const wrapper = mount(<PasswordInput />)

    expect(wrapper.find("input").length).toEqual(1)
  })

  it("renders masked by default", () => {
    const wrapper = mount(<PasswordInput />)

    expect(wrapper.find('input[type="password"]').length).toEqual(1)
    expect(wrapper.find('input[type="text"]').length).toEqual(0)
  })

  it("renders masked as a snapshot", () => {
    const component = renderer.create(<PasswordInput />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it("renders unmasked if you click on the eye", () => {
    const wrapper = mount(<PasswordInput />)

    const icon = wrapper.find("svg")
    icon.simulate("click")

    expect(wrapper.find('input[type="password"]').length).toEqual(0)
    expect(wrapper.find('input[type="text"]').length).toEqual(1)
  })

  it("renders unmasked as a snapshot", () => {
    const component = renderer.create(<PasswordInput />)

    const root = component.root
    const input = root.find(element => element.type === "svg")
    input.props.onClick()

    expect(component.toJSON()).toMatchSnapshot()
  })

  it("shows error instead of note when error exists", () => {
    const wrapper = mount(
      <PasswordInput note="This is a note" error="But this is an error" />
    )

    const text = wrapper.text()
    expect(text).not.toContain("This is a note")
    expect(text).toContain("But this is an error")
  })
})
