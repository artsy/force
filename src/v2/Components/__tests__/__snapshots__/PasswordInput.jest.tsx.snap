// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput renders masked as a snapshot 1`] = `
.c5 {
  position: relative;
}

.c0 {
  padding-bottom: 5px;
}

.c3 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 0;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 10px;
  line-height: initial;
}

.c1 {
  padding: 10px;
  box-shadow: none;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 10px;
  resize: none;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 1px solid #e5e5e5;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 0;
  margin-top: 5px;
  margin-bottom: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 20px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: content-box;
}

.c1::-webkit-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::-moz-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:-ms-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:hover,
.c1:focus,
.c1.focused {
  border-color: #6e1fff;
  outline: 0;
}

.c1:disabled {
  border: 2px dotted #e5e5e5;
}

.c2 {
  font-family: Unica77LLWebMedium,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  position: absolute;
  left: 10px;
  top: 10px;
  visibility: hidden;
  -webkit-animation: jiroXv 0.2s linear;
  animation: jiroXv 0.2s linear;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  z-index: 1;
}

.c4 {
  position: absolute;
  right: 10px;
  z-index: 1;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <label
      className="c2"
    />
    <input
      className="c3"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="password"
      value=""
    />
    <span
      className="c4"
    >
      <svg
        className="c5"
        fill="black100"
        height="18px"
        onClick={[Function]}
        viewBox="0 0 18 18"
        width="18px"
      >
        <title>
          View
        </title>
        <g
          fillRule="nonzero"
        >
          <path
            d="M2.249 8.462a.56.56 0 0 0 .016.601C4.172 11.9 6.406 13.286 9 13.286c2.594 0 4.828-1.387 6.736-4.225a.56.56 0 0 0 .017-.6C14.038 5.56 11.808 4.144 9.01 4.144S3.975 5.56 2.25 8.463zm-1.011-.574C3.163 4.652 5.77 3 9.01 3c3.241 0 5.843 1.653 7.755 4.89.331.561.31 1.257-.053 1.798-2.109 3.139-4.69 4.74-7.713 4.74-3.021 0-5.602-1.6-7.711-4.738a1.681 1.681 0 0 1-.05-1.802z"
          />
          <path
            d="M9 11.571a2.857 2.857 0 1 1 0-5.714 2.857 2.857 0 0 1 0 5.714zm0-1.142A1.714 1.714 0 1 0 9 7a1.714 1.714 0 0 0 0 3.429z"
          />
        </g>
      </svg>
    </span>
  </div>
</div>
`;

exports[`PasswordInput renders unmasked as a snapshot 1`] = `
.c5 {
  position: relative;
}

.c0 {
  padding-bottom: 5px;
}

.c3 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 0;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 10px;
  line-height: initial;
}

.c1 {
  padding: 10px;
  box-shadow: none;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 10px;
  resize: none;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 1px solid #e5e5e5;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 0;
  margin-top: 5px;
  margin-bottom: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 20px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: content-box;
}

.c1::-webkit-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::-moz-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:-ms-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:hover,
.c1:focus,
.c1.focused {
  border-color: #6e1fff;
  outline: 0;
}

.c1:disabled {
  border: 2px dotted #e5e5e5;
}

.c2 {
  font-family: Unica77LLWebMedium,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  position: absolute;
  left: 10px;
  top: 10px;
  visibility: hidden;
  -webkit-animation: jiroXv 0.2s linear;
  animation: jiroXv 0.2s linear;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  z-index: 1;
}

.c4 {
  position: absolute;
  right: 10px;
  z-index: 1;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <label
      className="c2"
    />
    <input
      className="c3"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value=""
    />
    <span
      className="c4"
    >
      <svg
        className="c5"
        fill="black100"
        height="18px"
        onClick={[Function]}
        viewBox="0 0 18 18"
        width="18px"
      >
        <title>
          Hide
        </title>
        <g
          fillRule="nonzero"
        >
          <path
            d="M2.249 8.605a.56.56 0 0 0 .016.6C4.172 12.044 6.406 13.43 9 13.43c2.594 0 4.828-1.387 6.736-4.225a.56.56 0 0 0 .017-.6c-1.714-2.901-3.944-4.318-6.742-4.318S3.975 5.703 2.25 8.605zM1.238 8.03C3.163 4.795 5.77 3.143 9.01 3.143c3.241 0 5.843 1.653 7.755 4.89.331.561.31 1.257-.053 1.798-2.109 3.139-4.69 4.74-7.713 4.74-3.021 0-5.602-1.6-7.711-4.738a1.681 1.681 0 0 1-.05-1.802z"
          />
          <path
            d="M9 11.714A2.857 2.857 0 1 1 9 6a2.857 2.857 0 0 1 0 5.714zm0-1.143a1.714 1.714 0 1 0 0-3.428 1.714 1.714 0 0 0 0 3.428z"
          />
          <path
            d="M15.054 2.053l.805.81-12.893 12.79-.805-.81z"
          />
        </g>
      </svg>
    </span>
  </div>
</div>
`;
